src/compression/mod.rs

```diff
@@ -0,0 +1,176 @@
+/// Module with functions and structs specific to brigde the las crate and laz crate to allow
+/// writing & reading LAZ data
+
+use std::io::{Write, Seek, Cursor, Read, SeekFrom};
+use {Header, Point, Result, Vlr};
+use writer::{PointWriter, write_point_to, write_header_and_vlrs_to};
+use reader::{PointReader, read_point_from};
+use laz::las::laszip::{LazVlr, LASZIP_DESCRIPTION, LASZIP_RECORD_ID, LASZIP_USER_ID};
+use std::fmt::Debug;
+
+
+pub(crate) fn create_laszip_vlr(laszip_vlr: &LazVlr) -> std::io::Result<Vlr> {
+    let mut cursor = Cursor::new(Vec::<u8>::new());
+    laszip_vlr.write_to(&mut cursor)?;
+    Ok(Vlr{
+        user_id: LASZIP_USER_ID.to_owned(),
+        record_id: LASZIP_RECORD_ID,
+        description: LASZIP_DESCRIPTION.to_owned(),
+        data: cursor.into_inner()
+    })
+}
+
+/// struct that knows how to decompress LAZ
+///
+/// Decompression is done in 2 step:
+/// 1) call the decompressor the read & decompress the next point
+/// and put its data in a in-memory buffer
+/// 2) read the buffer to get the decompress point
+pub(crate) struct CompressedPointReader<R: Read +Seek> {
+    /// decompressor that does the actual job
+    decompressor: laz::las::laszip::LasZipDecompressor<R>,
+    header: Header,
+    /// in-memory buffer where the decompressor writes deompression result
+    decomp_out: Cursor<Vec<u8>>,
+    last_point_idx: u64,
+}
+
+impl<R: Read + Seek> CompressedPointReader<R> {
+    pub(crate) fn new(source: R, mut header: Header) -> Result<Self> {
```

Comment:
Couple questions/comments here:

1. Can you explain why the vlr is removed from the header? Since the data are just passed by reference into `LazVlr::from_buffer`, we could just reference the data out of a header reference?
2. This might read cleaner if we add a helper method e.g. `Header::laz_vlr() -> Option<Vlr>` and move the finding logic into there?

Reply:
1) I decided to remove the VLR as it is really only useful for the Laszip reader, and if we decide to keep the vlr in the Header, when writing a new file, the `CompressedPointWriter` whould have to check if the header that was given already contains a laz_vlr, and if so, maybe check that it matches the point format.

2) I agree

Reply:
Thanks for the change on 2. For 1, I guess in my head it'd make more sense to do those checks and/or explicitly clear the laszip vlr on the writer side, since headers can be created from scratch and someone could create an unusable laszip vlr?